README.md

```diff
@@ -2,4 +2,61 @@
+
+Operators are assembled in a tree-like structure with every operator calling `Next()` on its dependants until there is no more data to be returned. The result of the `Next()` function is a _column vector_  (also called a _step vector_) with elements in the vector representing samples with the same timestamp from different time series.
+
+<p align="center">
+  <img src="./assets/design.png" />
+</p>
+
+This model allows for samples from individual time series to flow one execution step at a time from the left-most operators to the one at the very right. Since most PromQL expressions are aggregations, samples are reduced in number as they are pulled by the operators on the right. Because of this, samples from original timeseries can be decoded and kept in memory in batches instead of being fully expanded.
+
+In addition to operators that have a one-to-one mapping with PromQL constructs, the Volcano model also describes so-called Exchange operators which can be used for flow control and optimizations, such as concurrency or batched selects. An example of an _Exchange_ operator is described in the [Intra-operator parallelism](#intra-operator-parallelism) section.
+
+## Inter-operator parallelism
+
+Since operators are independent and rely on a common interface for pulling data, they can be run in parallel to each other. As soon as one operator has processed data from an evaluation step, it can pass the result onward so that its upstream can immediatelly start working on it.
+
+<p align="center">
+  <img src="./assets/promql-pipeline.png" />
+</p>
+
+
+## Intra-operator parallelism
+
+Parallelism can also be added within individual operators using a parallel coalesce exchange operator. Such exchange operators are indistinguishable from regular operators to their upstreams since they respect the same `Next()` interface.
+
+<p align="center">
+  <img src="./assets/parallel-coalesce.png" />
+</p>
+
+
+## Memory management 
+
+### Step vector allocations
+
+One challenge with the streamed execution model is knowing how much memory to allocate in each operator for each step. 
+
+To work around this issue, operators expose a `Series()` method which returns the labels for all time series that they will ever produce. Operators at the very bottom of the tree, like vector and matrix selectors, have this information since they are responsible for loading data from storage. Other operators can then call `Series()` on the downstream operator and pre-compute all possible outputs.
+
+Even thought this might look like an expensive operation, its cost is identical to just one evaluation step. Knowing sizes of input and output vectors also allows us to:   
+* Allocate memory very precisely by properly sizing vector pools (see section bellow),
+* Use arrays instead of maps for indexing data and performing lookups, leading to even further memory usage and faster execution time, and
```

Comment:
Does this mean that more memory is used, in lesser time?

Reply:
Ouch, this should say `leading to even lower memory usage`. Will change!